### PR TITLE
Optimize matchbin

### PR DESCRIPTION
**Pull request by mmore500**
Nibbling around the edges for about a 40% speedup on the tests.

### CODE REVIEW

**Comment by mmore500** [comment]
Running some random programs on SignalGP these changes seem to give ~5% speedup.

File: source/tools/matchbin_selectors.h [hunk 1]
```diff
@@ -138,11 +138,10 @@ namespace emp {
 
     RankedCacheState() = default;
     RankedCacheState(
-      emp::vector<size_t>::iterator begin,
-      emp::vector<size_t>::iterator back,
+      emp::vector<size_t> uids_,
       size_t n,
```
Comment (rodsan0) [inline source/tools/matchbin_selectors.h:142]: these can probably be const but idk if we care in this PR

**Review by rodsan0 — APPROVED** [review]
looks great! i see no reason not to merge this PR after the tests pass

**Comment by mmore500** [comment]
We should try `unordered_flat_map`

**Comment by mmore500** [comment]
Switching to flat map seems to improve performance maybe another 10%

**Pull request merged**